metomi/isodatetime/data.py

```diff
@@ -531,60 +646,76 @@ def __floordiv__(self, other):
                 "'%s' should be integer." %
                 type(other).__name__
             )
-        new = self.copy()
+        new = self._copy()
         if self.get_is_in_weeks():
-            new.weeks //= other
+            new._weeks //= other
             return new
-        new.years //= other
-        new.months //= other
-        new.days //= other
-        new.hours //= other
-        new.minutes //= other
-        new.seconds //= other
+        new._years //= other
+        new._months //= other
+        new._days //= other
+        new._hours //= other
+        new._minutes //= other
+        new._seconds //= other
         return new
 
-    def __eq__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data == other_data
+    def __hash__(self) -> int:
+        # TODO: alt calendar modes
+        if self.get_is_in_weeks():
+            return hash((0, 0, self._get_non_nominal_seconds()))
+        return hash(
+            (self._years, self._months, self._get_non_nominal_seconds()))
 
-    def __ne__(self, other: "Duration") -> bool:
-        return not self.__eq__(other)
+    def __eq__(self, other: "Duration") -> bool:
+        if isinstance(other, Duration):
+            if self.is_exact():
+                if other.is_exact():
+                    return (self._get_non_nominal_seconds() ==
+                            other._get_non_nominal_seconds())
+                return False
+            return (
+                self._years == other._years and
+                self._months == other._months and
+                self._get_non_nominal_seconds() ==
+                other._get_non_nominal_seconds()
+            )
+        return NotImplemented
```

Comment:
Might be good to `return False` here otherwise we will have problems comparing to `None` instances.

Ok to raise exceptions in other cases though.

Reply:
Returning `NotImplemented` delegates to `other`'s `__eq__` method, which for `None` seems to do something along the lines of
```python
return False if other is not None else True
```
So shouldn't be a problem

Reply:
Sorry, I read `raise NotImplementedError` rather than `return NotImplemented`!